### DISCUSSION TITLE
[custom cache middleware] How to produce a response from cache but allow the request to continue?

### DISCUSSION THREAD

**Discussion by lilouartz** [fastify/fastify#5644] [General]
I am building a cache middleware for my website:

```ts
app.addHook('onRequest', async (request, reply) => {
  if (request.method !== 'GET') {
    return;
  }

  if (request.url.includes('/api/')) {
    return;
  }

  // Do not attempt to use cache for authenticated visitors.
  if (request.visitor.userAccount) {
    return;
  }

  const cacheKey = generateCacheKey(request);

  if (cacheKey) {
    const cachedResponse = await redis.get(cacheKey);

    if (cachedResponse) {
      reply.header('x-pillser-cache', 'HIT');
      reply.send(cachedResponse);
      reply.hijack();

      return;
    }
  }

  reply.header('x-pillser-cache', 'MISS');
});
```

The challenge is that I want cache to "refresh" in the background even if I produce `HIT` response. 

Effectively, I am trying to implement `stale-while-revalidate`.

How can I do that?

**Comment by kilisamemarisaaa** [comment]
`reply.send()` is the terminal action for this request. Once a hook sends the reply, Fastify skips the remaining hooks and route handler; `reply.hijack()` is not needed after `send()` and cannot make the original handler continue safely. Continuing that handler would eventually attempt a second response.

For stale-while-revalidate, separate producing fresh content from sending the current reply. Start that work independently, then return the cached response:

```js
const refreshes = new Map()

function revalidateOnce (cacheKey, input, log) {
  if (refreshes.has(cacheKey)) return

  const task = buildFreshResponse(input)
    .then((fresh) => redis.set(cacheKey, fresh, { EX: CACHE_TTL_SECONDS }))
    .catch((err) => log.error({ err, cacheKey }, 'cache revalidation failed'))
    .finally(() => refreshes.delete(cacheKey))

  refreshes.set(cacheKey, task)
}

app.addHook('onRequest', async (request, reply) => {
  // ...your bypass checks...

  const cacheKey = generateCacheKey(request)
  const cachedResponse = cacheKey && await redis.get(cacheKey)

  if (cachedResponse) {
    // Pass only the data needed to regenerate; do not reuse this request's reply.
    revalidateOnce(cacheKey, snapshotInput(request), request.log)

    return reply
      .header('x-pillser-cache', 'HIT')
      .send(cachedResponse)
  }

  reply.header('x-pillser-cache', 'MISS')
})
```

The per-key `refreshes` map is important: without single-flight protection, every stale hit can start the same regeneration and cause a cache stampede. In a multi-process deployment, use a Redis lock or queue instead of an in-memory map.

If extracting `buildFreshResponse()` is not practical, a background `app.inject()` request with an explicit cache-bypass marker can reuse the complete route, but it must prevent recursion and deliberately reproduce the required authentication/request context. A dedicated function or job is usually safer.

**Comment by usualdesigner** [comment]
Since you already control what happens after reply.hijack(), that's where stale-while-revalidate goes - you've sent the cached response and taken over the reply, so nothing stops you kicking off the revalidation as fire-and-forget right after. Just don't await it: call your regenerate-and-cache function without awaiting, let it run in the background, return. The response has already gone out, there's no reply object left to interfere with.
